File: ChelaVm/src/Class.cpp
#include "ChelaVm/Class.hpp"

namespace ChelaVm
{
	Class::Class(Module *module)
		: Structure(module)
	{
	}
	
	Class::~Class()
	{
	}
		
	bool Class::IsStructure() const
	{
		return false;
	}
	
	bool Class::IsClass() const
	{
		return true;
	}

    bool Class::IsInterface() const
    {
        return false;
    }

    bool Class::IsPassedByReference() const
    {
        return true;
    }

    Class *Class::PreloadMember(Module *module, ModuleReader &reader, const MemberHeader &header)
    {
        // Create the result.
        Class *res = new Class(module);

        // Store the name and the member flags.
        res->name = module->GetString(header.memberName);
        res->flags = (MemberFlags)header.memberFlags;

        // Skip the member data.
        reader.Skip(header.memberSize);
        return res;
    }

    Structure *Class::CreateBuilding(Module *module) const
    {
        return new Class(module);
    }
};

File: ChelaCompiler/Module/Class.cs
using Chela.Compiler.Ast;

namespace Chela.Compiler.Module
{
	public class Class: Structure
	{
        protected Class(ChelaModule module)
            : base(module)
        {
        }

		public Class(string name, MemberFlags flags, Scope parentScope)
			: base(name, flags, parentScope)
		{
		}
		
		public override bool IsStructure()
		{
			return false;
		}
		
		public override bool IsClass ()
		{
			return true;
		}

        public override bool IsPassedByReference ()
        {
            return true;
        }

        public bool IsPartial()
        {
            return (GetFlags() & MemberFlags.ImplFlagMask) == MemberFlags.Partial;
        }

        internal new static void PreloadMember(ChelaModule module, ModuleReader reader, MemberHeader header)
        {
            // Create the temporal structure and register it.
            Class clazz = new Class(module);
            module.RegisterMember(clazz);

            // Read the name.
            clazz.name = module.GetString(header.memberName);
            clazz.flags = (MemberFlags)header.memberFlags;

            // Skip the class elements.
            reader.Skip(header.memberSize);
        }
	}
}

